File: crossref.py
#!/usr/bin/python2.7

import urllib
import urllib2
import xml.etree.ElementTree

username = 'bgamari@physics.umass.edu'
password = 'mudpie11'
doi_url = 'http://doi.crossref.org/servlet/query'

class CrossRefError(RuntimeError):
        def __init__(self, text):
                self.text = text
        def __str__(self):
                return self.text

def _parse_journal_record(journal_rec):
        et = journal_rec
        d = {}
        def set_key(key, val):
                if val: d[key] = val
        d['type'] = 'journal_article'
        d['doi'] = et.findtext('journal_article/doi_data/doi')
        set_key('journal', et.findtext('journal_metadata/abbrev_title'))
        set_key('full_journal', et.findtext('journal_metadata/full_title'))
        set_key('volume', et.findtext('journal_issue/journal_volume/volume'))
        set_key('issue', et.findtext('journal_issue/issue'))
        set_key('title', et.findtext('journal_article/titles/title'))
        set_key('year', et.findtext('journal_article/publication_date/year'))
        authors = et.findall('journal_article/contributors/person_name')
        if authors:
                d['authors'] = [ {'forenames': a.findtext('given_name'),
                                  'surname': a.findtext('surname')} for a in authors]

        first = et.findtext('journal_article/pages/first_page')
        last = et.findtext('journal_article/pages/last_page')
        if first and last:
                d['pages'] = '%s-%s' % (first, last)
        elif first:
                d['pages'] = first
                                  
        return d

def lookup_doi(doi):
        url = '%s?usr=%s&pwd=%s&format=xml&id=%s' % (doi_url, username, password, urllib.quote(doi))
        f = urllib2.urlopen(url)
        d = f.read()
        et = xml.etree.ElementTree.fromstring(d)
        err = et.findtext('doi_record/crossref/error')
        if err is not None:
                raise CrossRefError(err)

        j = et.find('doi_record/crossref/journal')
        if j is not None:
                return _parse_journal_record(j)
        else:
                return None

if __name__ == '__main__':
        import sys
        dois = sys.argv[1:]
        if len(dois) == 0:
                dois = ['10.1577/H02-043', '10.1021/jp035514+']
        for doi in dois:
                print lookup_doi(doi)

File: RefTrack/CrossRef.hs
{-# LANGUAGE OverloadedStrings #-}

module RefTrack.CrossRef where

import qualified Data.ByteString.Lazy.Char8 as LBS

import           Control.Applicative
import           Control.Error.Safe       
import           Data.Maybe
import           Data.List
import qualified Data.Set as S
import           Data.Text (Text)
import qualified Data.Text as T
import           Control.Monad
import           Control.Monad.Trans.Class
import           Control.Monad.Trans.Error
import           Data.Functor.Identity
import           Network.HTTP
import           Network.URI
import           RefTrack.Types hiding (Document)
import           Text.XML
import           Text.XML.Cursor

userName = "bgamari@physics.umass.edu"
passwd = "mudpie11"
hostname = URIAuth { uriUserInfo = ""
                   , uriRegName = "doi.crossref.org"
                   , uriPort = ""
                   }

buildQuery :: [(String,String)] -> String
buildQuery q = "?"++(intercalate "&" $ map (\(k,v)->k++"="++v) q)
               
liftEither :: (Monad m, Error e') => (e -> e') -> Either e a -> ErrorT e' m a
liftEither f (Left e) = throwError $ f e
liftEither f (Right a) = return a

mapErr ::  (Monad m, Error e') => (e -> e') -> ErrorT e m a -> ErrorT e' m a
mapErr f = mapErrorT (>>= g)
    where g (Left e)  = return $ Left $ f e
          g (Right a) = return $ Right a
                 
lookupDoi :: DOI -> ErrorT String IO (Maybe Ref)
lookupDoi (DOI doi) = do
    let uri = URI { uriScheme = "http:"
                  , uriAuthority = Just hostname
                  , uriPath = "/servlet/query"
                  , uriQuery = buildQuery [ ("usr", userName)
                                          , ("pwd", passwd)
                                          , ("format", "xml")
                                          , ("id", normalizeEscape $ T.unpack doi)
                                          ]
                  , uriFragment = ""
                  }
    resp <- liftEither show =<< (lift $ simpleHTTP $ mkRequest GET uri)
    case resp of
      Response {rspCode=(2,_,_), rspBody=body} -> do
        doc <- liftEither show $ parseLBS def body
        let queries = fromDocument doc
                      $/ laxElement "query_result"
                      &/ laxElement "body"
                      &/ laxElement "query"
                      >=> attributeIs "status" "resolved"
        case queries of
          r:_ -> do a <- parseQueryReply r
                    return $ Just a
          []  -> throwError "No records returned"
      a -> throwError $ show a
       
ensureNotNull :: (Monad m, Error e) => e -> [a] -> ErrorT e m a
ensureNotNull e [] = throwError e
ensureNotNull _ (a:_) = return a

parseQueryReply :: Monad m => Cursor -> ErrorT String m Ref
parseQueryReply root = do
    let e = root $/ laxElement "crossref"
                 &/ laxElement "error"
    when (not $ null e) $ throwError $ T.unpack $ head $ content $ head e
    
    j <- ensureNotNull "Can't find journal element"
         $ root $/ laxElement "crossref"
                &/ laxElement "journal"
    ja <- ensureNotNull "Can't find journal_article element"
          $ j $/ laxElement "journal_article"
    ji <- ensureNotNull "Can't find journal_issue element"
          $ j $/ laxElement "journal_issue" 
    let vol = maybe 0 id $ do jv <- listToMaybe $ ji $/ laxElement "journal_volume"
                                                     &/ laxElement "volume"
                                                     &// content
                              readZ $ T.unpack jv
    let pub = JournalIssue
              { _pubVolume = vol
              , _pubFullTitle = T.concat $ j $/ laxElement "journal_metadata"
                                             &/ laxElement "full_title"
                                             &/ content
              , _pubAbbrevTitle = listToMaybe $ j $/ laxElement "journal_metadata"
                                                  &/ laxElement "abbrev_title"
                                                  &/ content
              , _pubIssue = T.concat $ j $/ laxElement "journal_volume"
                                         &/ laxElement "issue"
                                         &/ content
              }

    let parsePerson e = Person <$> (listToMaybe $ e $/ laxElement "given_name" &/ content)
                               <*> (listToMaybe $ e $/ laxElement "surname" &/ content)
        authors = mapMaybe parsePerson
                  $ ja $/ laxElement "contributors"
                       &/ laxElement "person_name"
                       >=> attributeIs "contributor_role" "author"
    
    let erefs = map (DOIRef . DOI) $ ja $/ laxElement "doi_data"
                                        &/ laxElement "doi"
                                        &/ content
    let pubYear = do year <- listToMaybe $ ja $/ laxElement "publication_date"
                                              &/ laxElement "year"
                                              &/ content
                     readZ $ T.unpack year 
    return $ Ref { _refId = RefId ""
                 , _refTitle = T.strip $ T.intercalate " " $ map (T.unwords . T.words)
                               $ ja $/ laxElement "titles" &// content
                 , _refAuthors = authors
                 , _refPubDate = Nothing
                 , _refYear = Year $ maybe 0 id pubYear
                 , _refExternalRefs = erefs
                 , _refPublication = pub
                 , _refAbstract = Nothing
                 , _refTags = S.empty
                 }
              
parseDoiRecord :: Cursor -> Ref
parseDoiRecord cursor =  undefined
